File: CPlusPlus/Sample_Source/Text/ExtractTextPreservingStyleAndPositionInfo/ExtractTextPreservingStyleAndPositionInfo.cpp
//
// Copyright (c) 2023, Datalogics, Inc. All rights reserved.
//
// For complete copyright information, refer to:
// http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/
//
// This sample extracts text, quads, and styles from a PDF
// document and saves the text to a file.
//
// For more detail see the description of the ExtractTextPreservingStyleAndPositionInfo
// sample program on our Developer’s site,
// http://dev.datalogics.com/adobe-pdf-library/sample-program-descriptions/c1samples

#include <fstream>
#include <string>
#include <sstream>

#include "InitializeLibrary.h"
#include "APDFLDoc.h"
#include "TextExtract.h"

#ifdef MAC_PLATFORM
#include "../../_Common/ThirdParty/json.hpp"
#else
#include "ThirdParty/json.hpp"
#endif

// Set the namespace for the JSON library
using json = nlohmann::json;

const char *DEF_INPUT = "../../../../Resources/Sample_Input/sample.pdf";
const char *DEF_OUTPUT = "ExtractTextPreservingStyleAndPositionInfo-out.json";

void SaveJson(std::vector<PDTextAndDetailsExtractRec> extractedText);

int main(int argc, char **argv) {

    // Initialize the library
    APDFLib libInit;
    ASErrorCode errCode = 0;
    if (libInit.isValid() == false) {
        errCode = libInit.getInitError();
        std::cout << "Initialization failed with code " << errCode << std::endl;
        return libInit.getInitError();
    }

    DURING

        APDFLDoc inAPDoc(DEF_INPUT, true);

        TextExtract textExtract(inAPDoc.getPDDoc());

        std::vector<PDTextAndDetailsExtractRec> extractedText = textExtract.GetTextAndDetails();

        SaveJson(extractedText);

    HANDLER
        errCode = ERRORCODE;
        libInit.displayError(errCode);
    END_HANDLER

    return errCode;
}

//  Helper function to set decimal precision in JSON file
std::string to_string_with_precision(float a_value, const int n = 2)
{
    std::ostringstream out;
    out.precision(n);
    out << std::fixed << a_value;
    return out.str();
}

// Create JSON file of results
void SaveJson(std::vector<PDTextAndDetailsExtractRec> extractedText)
{
    // This array will hold the JSON stream that we will print to the output JSON file.
    json result = json::array();

    for (size_t textIndex = 0; textIndex < extractedText.size(); ++textIndex) {

        json textObject = json::object();
        textObject["text"] = extractedText[textIndex].text;
        for (size_t quadIndex = 0; quadIndex < extractedText[textIndex].boundingQuads.size(); ++quadIndex) {
            json quadPointValues = json::array();
            quadPointValues.push_back(to_string_with_precision(extractedText[textIndex].boundingQuads[quadIndex].tl.h));
            quadPointValues.push_back(to_string_with_precision(extractedText[textIndex].boundingQuads[quadIndex].tl.v));
            textObject["quad"]["top-left"] = quadPointValues;
            quadPointValues.clear();

            quadPointValues.push_back(to_string_with_precision(extractedText[textIndex].boundingQuads[quadIndex].tr.h));
            quadPointValues.push_back(to_string_with_precision(extractedText[textIndex].boundingQuads[quadIndex].tr.v));
            textObject["quad"]["top-right"] = quadPointValues;
            quadPointValues.clear();

            quadPointValues.push_back(to_string_with_precision(extractedText[textIndex].boundingQuads[quadIndex].bl.h));
            quadPointValues.push_back(to_string_with_precision(extractedText[textIndex].boundingQuads[quadIndex].bl.v));
            textObject["quad"]["bottom-left"] = quadPointValues;
            quadPointValues.clear();

            quadPointValues.push_back(to_string_with_precision(extractedText[textIndex].boundingQuads[quadIndex].br.h));
            quadPointValues.push_back(to_string_with_precision(extractedText[textIndex].boundingQuads[quadIndex].br.v));
            textObject["quad"]["bottom-right"] = quadPointValues;
            quadPointValues.clear();
        }
        json styles = json::array();
        for (size_t styleIndex = 0; styleIndex < extractedText[textIndex].styles.size(); ++styleIndex) {
            json styleObject = json::object();
            styleObject["char-index"] = extractedText[textIndex].styles[styleIndex].charIndex;
            styleObject["font-name"] = extractedText[textIndex].styles[styleIndex].fontname;
            styleObject["font-size"] = to_string_with_precision(extractedText[textIndex].styles[styleIndex].fontsize);
            styleObject["color-space"] = extractedText[textIndex].styles[styleIndex].colorValues.DLSpace;
            json colorValues = json::array();
            for (size_t colorIndex = 0; colorIndex < 4; ++colorIndex) {
                colorValues.push_back(extractedText[textIndex].styles[styleIndex].colorValues.DLColor[colorIndex]);
            }
            styleObject["color-values"] = colorValues;
            styles.push_back(styleObject);
        }
        textObject["styles"] = styles;
        result.push_back(textObject);
    }
    std::ofstream jsonOutputFileStream(DEF_OUTPUT, std::ios_base::trunc | std::ios_base::out);
    jsonOutputFileStream << std::setw(4) << result;
    jsonOutputFileStream.close();
}

File: DotNET/Sample_Source/Text/ExtractTextPreservingStyleAndPositionInfo/ExtractTextPreservingStyleAndPositionInfo.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.IO;
using Datalogics.PDFL;
using ExtractTextNameSpace;

/*
 * 
 * This sample extracts text and details of that text in a PDF
 * document, prints to console, and saves the text to a JSON file.
 *
 * Copyright (c) 2022-2023, Datalogics, Inc. All rights reserved.
 *
 * For complete copyright information, refer to:
 * http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/.
 *
 */
namespace ExtractTextPreservingStyleAndPositionInfo
{
    class ExtractTextPreservingStyleAndPositionInfo
    {
        // Set Defaults
        static String sInput = Library.ResourceDirectory + "Sample_Input/sample.pdf";
        static String sOutput = "../ExtractTextPreservingStyleAndPositionInfo-out.json";

        static void Main(string[] args)
        {
            Console.WriteLine("ExtractTextPreservingStyleAndPositionInfo Sample:");


            using (new Library())
            {
                Console.WriteLine("Initialized the library.");

                Console.WriteLine("Input file: " + sInput);

                using (Document doc = new Document(sInput))
                {
                    using (ExtractText docText = new ExtractText(doc))
                    {
                        List<TextAndDetailsObject> result = docText.GetTextAndDetails();

                        // Save the output to a JSON file.
                        SaveJson(result);
                    }
                }
            }
        }

        static void SaveJson(List<TextAndDetailsObject> result)
        {
            using FileStream fs = File.Create(sOutput);
            var writerOptions = new JsonWriterOptions { Indented = true };
            using (var writer = new Utf8JsonWriter(fs, options: writerOptions))
            {
                writer.WriteStartArray();
                foreach (TextAndDetailsObject resultText in result)
                {
                    writer.WriteStartObject();
                    writer.WriteString("text", resultText.Text);
                    writer.WriteStartArray("quads");
                    foreach (Quad quad in resultText.Quads)
                    {
                        writer.WriteStartObject();
                        writer.WriteString("top-left", quad.TopLeft.ToString());
                        writer.WriteString("top-right", quad.TopRight.ToString());
                        writer.WriteString("bottom-left", quad.BottomLeft.ToString());
                        writer.WriteString("bottom-right", quad.BottomRight.ToString());
                        writer.WriteEndObject();
                    }
                    writer.WriteEndArray();
                    writer.WriteStartArray("styles");
                    foreach (DLStyleTransition st in resultText.StyleList)
                    {
                        writer.WriteStartObject();
                        writer.WriteString("char-index", st.CharIndex.ToString());
                        writer.WriteString("font-name", st.Style.FontName.ToString());

                        writer.WriteString("font-size", Math.Round(st.Style.FontSize, 2).ToString());
                        writer.WriteString("color-space", st.Style.Color.Space.Name);
                        writer.WriteStartArray("color-values");
                        foreach (double cv in st.Style.Color.Value)
                        {
                            writer.WriteStringValue(Math.Round(cv, 3).ToString());
                        }
                        writer.WriteEndArray();
                        writer.WriteEndObject();
                    }
                    writer.WriteEndArray();
                    writer.WriteEndObject();
                }
                writer.WriteEndArray();
                writer.Flush();
            }
        }
    }
}
